### PR TITLE
Implement 'deletePortfolioDraft' function

### PR DESCRIPTION
**Pull request by jasonburkert-ccpo**
Implements `deletePortfolioDraft` function for AWS.

Deletes entire Portfolio Draft item from DynamoDB.
Follows #44 as an example and leverages responses and errors from #46.

Ticket: AT-6364

### CODE REVIEW

**Comment by jasonburkert-ccpo** [comment]
Camelcase config mirroring 62bf70dbd5a17adae540a68a43b0abd02ab06f1a

**Comment by jasonburkert-ccpo** [comment]
Scripts mirror those in `./infrastructure/package.json`

**Comment by jasonburkert-ccpo** [comment]
Happy-path working.

File: infrastructure/lib/atat-web-api-stack.ts
Comment (kylelaker-ccpo) [inline infrastructure/lib/atat-web-api-stack.ts:97]:
```suggestion
    const portfolioDraft = portfolioDrafts.addResource("{portfolioDraftId}");
```

Maybe since once we've used the `{portfolioDraftID}` we're dealing with _one_ portfolio? We could even call these `portfolioDraftsRoute` and `portfolioDraftRoute` if that helps?

File: applications/portfolioDrafts/package.json
Comment (kylelaker-ccpo) [inline applications/portfolioDrafts/package.json:15]: Do we need `@aws-sdk/util-dynamodb` for something ourselves? Is this a peer dependency? I have no problem with adding it; I just want to make sure I understand for future applications

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (kylelaker-ccpo) [inline applications/portfolioDrafts/deletePortfolioDraft.ts:16]:
```suggestion
  if (!portfolioDraftId) {
```

To catch `undefined` or `""` as well?

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (kylelaker-ccpo) [inline applications/portfolioDrafts/deletePortfolioDraft.ts:17]: We should talk about defining an error result type.

```
{
  "errorCode": "InvalidRequest",
  "errorMessage": "A valid portfolio draft ID must be specified"
}
```

And maybe some helper functions? Eventually, it may be useful for the front end to parse info about errors so we'll want an expandable format for that.

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (kylelaker-ccpo) [inline applications/portfolioDrafts/deletePortfolioDraft.ts:22]: I am not sure that I like exposing the implementation detail that an ID must be a valid UUID. I think we can still perform this check but maybe return a `404` and a say that a Portfolio Draft with that ID doesn't exist (because we know it won't)?

File: .eslintrc.json
Comment (kylelaker-ccpo) [inline .eslintrc.json:26]: So the only place we really want to allow non-camel-case is in the models, right? Because those are required to be snake case per the API specification? Thoughts on just exempting `models/**.ts` in the `overrides` section below? Something like this?
```
{
  "files": ["models/**.ts"],
  "rules": { "camelcase": "off" }
}
```

I'd like to avoid letting snake case sneak into the rest of the app code or the infrastructure. And we can exempt single lines as needed with `eslint-disable-line camelcase` later if we need to?

**Review by kylelaker-ccpo — CHANGES_REQUESTED** [review]
Great to see this added! Thank you! I know this is still in draft but I did add a few questions to try to get them answered earlier in the dev process if possible

File: infrastructure/lib/atat-web-api-stack.ts
Comment (jasonburkert-ccpo) [reply infrastructure/lib/atat-web-api-stack.ts:97]: Thanks for looking at this.  I was pulled into meetings this afternoon.

**Comment by zachclark-ccpo** [comment]
Working on changing the response, because we are receiving the metadata instead of the proper `DeleteCommandOutput`
```
{
  "$metadata": {
    "httpStatusCode": 200,
    "requestId": "OSQ1GI6APMQDOA11HPSMLLCCBJVV4KQNSO5AEMVJF66Q9ASUAAJG",
    "attempts": 1,
    "totalRetryDelay": 0
  }
}
```

File: applications/portfolioDrafts/package.json
Comment (kylelaker-ccpo) [reply applications/portfolioDrafts/package.json:15]: It's a peer dependency

Comment (zachclark-ccpo) [reply applications/portfolioDrafts/package.json:15]: I'll make a commit to remove it, its a peer dependency of `"node_modules/@aws-sdk/lib-dynamodb"`

File: .eslintrc.json
Comment (jasonburkert-ccpo) [reply .eslintrc.json:26]: I'm going to reverse this and deal with it later.   I was trying to avoid having the linter suggest too many changes related to camelCase.  But that doesn't really belong in this PR with the function impl.

Comment (jasonburkert-ccpo) [reply .eslintrc.json]: Reversed with https://github.com/dod-ccpo/atat-web-api/pull/40/commits/ecf1410cd1c61bb71bffe8dbdba0dadcf228ab9b

**Comment by jasonburkert-ccpo** [comment]
Working nicely at the expected route, issuing status 204 using class `NoContentResponse`.

image

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (kylelaker-ccpo) [inline applications/portfolioDrafts/deletePortfolioDraft.ts:49]: Does the constructor require this? I was expecting to work like:

```
return new NoContentResponse()
```

Since there should not be a body and the status code should be 204.

**Review by jeffsegal-ccpo — CHANGES_REQUESTED** [review]
Needs a UT

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (kylelaker-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts:49]: As-is, I think this will create a _header_ that looks like:

```
Message: Successful operation
```

That probably won't break anything but I think we should just go with an empty constructor?

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (zachclark-ccpo) [inline applications/portfolioDrafts/deletePortfolioDraft.ts:50]:
```suggestion
  } catch (err) {
  console.log("Database error: " + err);
```
We should log the database error, to keep it consistent with the other fns.

Comment (kylelaker-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts:50]: @zachclark-ccpo I agree that is correct _for now_. I'd really like to see more abstractions around our DB interactions so that we don't have to worry about "manually" logging quite so much. We should see if we can get a ticket in the next sprint to support that.

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (jeffsegal-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts:49]: I realized that I didn't do the delete responses the way I had intended (200 should return the entity being deleted and 204 should return an empty body). We can fix that in another PR though. 

And to Kyle's point - an empty constructor would be preferable to passing boilerplate JSON into the constructor.

**Comment by zachclark-ccpo** [comment]
> Needs a UT

Should we do a separate PR for testing framework?

File: infrastructure/lib/atat-web-api-stack.ts
Comment (jeffsegal-ccpo) [inline infrastructure/lib/atat-web-api-stack.ts:100]: Can we add a TODO to reduce the boilerplate here? Perhaps we can pass a little config object into a method like this?

```
{
  myNodejsFunction,
  parentRoute,
  "POST",
  table.grantReadWriteData
}
```

Feel free to tell me that I'm a dumb Java guy but I figured we can do something that helps us avoid the dreaded copypasta that can come with a long list of named variables and calling different methods that will all compile but not behave properly.

@kylelaker-ccpo this also brings up the question of how much value we might get out of applying a UT to our stacks here.

**Review by jeffsegal-ccpo — CHANGES_REQUESTED** [review]

**Comment by jeffsegal-ccpo** [comment]
> > Needs a UT
> 
> Should we do a separate PR for testing framework?

I'd really like to avoid the whole thing where we put off testing until later. I'd much rather see a simple UT that we have to optimize later so that we're not letting the perfect be the enemy of the good.

File: infrastructure/lib/atat-web-api-stack.ts
Comment (kylelaker-ccpo) [reply infrastructure/lib/atat-web-api-stack.ts:100]: Absolutely! I am actually planning to work on that today/tomorrow as well. I want to have a `registerDynamoDbInteractionFunction` or something like that with a less awful name that you just pass the API GW resource, the file name, the table, and the HTTP method to and it does some magic for you. The big comment at the top of this section is basically a `// TODO: Reduce copypasta`.

As far as how to test the infra code, working on that too. We should do it. We just need to think about what parts we _actually_ want to test here.

There's also the fact that API GW can "import" an OpenAPI spec, but I don't think that's elegant in the CDK (or CloudFormation generally) today since it's not easy to reference the Lambda integrations, etc. But I am sorta playing around with that idea a bit (but not in any form that will make it to a PR soon).

**Comment by kylelaker-ccpo** [comment]
> > > Needs a UT
> > 
> > 
> > Should we do a separate PR for testing framework?
> 
> I'd really like to avoid the whole thing where we put off testing until later. I'd much rather see a simple UT that we have to optimize later so that we're not letting the perfect be the enemy of the good.

I think the issue is that we may not really have a lot of the "plumbing" for tests in place yet. If that does need to be done, the plumbing needs to be in another PR and then this PR needs to be rebased to consume that plumbing. I'd love to be wrong here though

**Comment by jeffsegal-ccpo** [comment]
> > > > Needs a UT
> > > 
> > > 
> > > Should we do a separate PR for testing framework?
> > 
> > 
> > I'd really like to avoid the whole thing where we put off testing until later. I'd much rather see a simple UT that we have to optimize later so that we're not letting the perfect be the enemy of the good.
> 
> I think the issue is that we may not really have a lot of the "plumbing" for tests in place yet. If that does need to be done, the plumbing needs to be in another PR and then this PR needs to be rebased to consume that plumbing. I'd love to be wrong here though

That's fair. I don't want to hold this one up if we don't even have the plumbing in place. But @jasonburkert-ccpo can you create a new ticket for A) creating said plumbing and B) applying it to all existing routes? Once that merges we'll make it a baseline requirement for all new PRs to include tests.

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (jasonburkert-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts:49]: I was trying to provide the description: Successful operation specified for the 200 in the API, but was just clumsy.  Will remove.

Comment (jasonburkert-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts]: Fixed in https://github.com/dod-ccpo/atat-web-api/pull/40/commits/b43d75a429b5d25a08a140d4018707d5828282c7

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (jasonburkert-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts:50]: Fixed in https://github.com/dod-ccpo/atat-web-api/pull/40/commits/6a7390025deab72285d0bcc32a9b31cabecb6011

**Review by jeffsegal-ccpo — APPROVED** [review]

File: infrastructure/lib/atat-web-api-stack.ts
Comment (jasonburkert-ccpo) [reply infrastructure/lib/atat-web-api-stack.ts:100]: I understood that we wish to:

- reduce boilerplate in `infrastructure/lib/atat-web-api-stack.ts`
- add unit testing against software stack infrastructure

**Comment by jasonburkert-ccpo** [comment]
> > > > > Needs a UT
> > > > 
> > > > 
> > > > Should we do a separate PR for testing framework?
> > > 
> > > 
> > > I'd really like to avoid the whole thing where we put off testing until later. I'd much rather see a simple UT that we have to optimize later so that we're not letting the perfect be the enemy of the good.
> > 
> > 
> > I think the issue is that we may not really have a lot of the "plumbing" for tests in place yet. If that does need to be done, the plumbing needs to be in another PR and then this PR needs to be rebased to consume that plumbing. I'd love to be wrong here though
> 
> That's fair. I don't want to hold this one up if we don't even have the plumbing in place. But @jasonburkert-ccpo can you create a new ticket for A) creating said plumbing and B) applying it to all existing routes? Once that merges we'll make it a baseline requirement for all new PRs to include tests.

Created ticket AT-6456 for unit test plumbing.

**Comment by jasonburkert-ccpo** [comment]
Change requests have been addressed.  Ready for approve and merge.

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (kylelaker-ccpo) [inline applications/portfolioDrafts/deletePortfolioDraft.ts:29]: Do we actually want to _reject_ requests with a body? Or just choose to ignore them? I can see an argument either way. The specification is unclear as it doesn't define a body (where as `POST /portfolioDrafts` explicitly specifies it _must_ be empty). Should we decide on a convention here?

I like the strategy of being liberal in what we accept and conservative in what we emit; so with that since the spec doesn't specify a body schema, we should either:
- update the spec to forbid non-empty bodies
- accept and ignore the body

And we may want to consider the possibility that we _could_ actually accept parameters to the `DELETE` operation in the future.

Comment (jasonburkert-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts:29]: I could be convinced that we should be liberal in what we accept.  The DELETE in the spec does not specify a requestBody.

**Review by kylelaker-ccpo — APPROVED** [review]
My previous question is not a blocker. I am going to approve this with the expectation that we _do_ have the conversation on request bodies soon

File: applications/portfolioDrafts/deletePortfolioDraft.ts
Comment (jasonburkert-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts:29]: I think it's just copy-paste from `POST /portfolioDrafts`.  Removing.

Comment (jasonburkert-ccpo) [reply applications/portfolioDrafts/deletePortfolioDraft.ts]: Fix in https://github.com/dod-ccpo/atat-web-api/pull/40/commits/c9487c1f822a01d4b945c5165b237515be2bc6ec

**Pull request merged**
